### PR TITLE
Add support to allow routing to the weighted cluster specified in the request_header

### PR DESCRIPTION
**Pull request by tyxia**
The proto API is approved and merged here https://github.com/envoyproxy/envoy/pull/17673.

- Add `pickWeightedCluster` function as member function of `RouteEntryImplBase`. It is because 1)currently, `WeightedClusterUtil::pickCluster(..)` is templated free function shared with tcp_proxy whose weighted clusters don't /shouldn't have the `cluster_header`. 2) Here we need to get access to `DynamicRouteEntry` which is private of `RouteEntryImplBase`

- Refactor the code to put cluster_header pick logic into single function `pickClusterViaClusterHeader` that can be reused. 

- Add custom filter + weighted cluster integration test to simulate the behavior that : 1) PickCluster 2) ClearRouteCache and modify the header to map the cluster header to cluster name 3) re-pick the cluster at router filter.

- Modify `sendRequestAndWaitForResponse` interface to let it take in a list of upstream indices.

- Refactor the base integration test framework and `MockRandomGenerator` to take in fixed RNG value that could be specified by the test (other than just 0). This make the cluster selection (hence the test) deterministic and add flexibility to the test framework.

Signed-off-by: Tianyu Xia <tyxia@google.com>

Risk Level: Low
Testing:  CI, newly added integration test and unit test
ReleaseNote: Added

### CODE REVIEW

**Review by adisuissa — COMMENTED** [review]
From API standpoint - this PR is waiting for #17673 to be merged, so no need to review it at the moment.
I'll review the code next.

File: test/integration/weighted_cluster_integration_test.cc [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        [this, &total_cluster_with_name_count, &weights](
+            envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+                hcm) {
+          for (int i = 0; i < total_cluster_with_name_count; ++i) {
+            auto* cluster_with_name = hcm.mutable_route_config()
+                                          ->mutable_virtual_hosts(0)
+                                          ->mutable_routes(0)
+                                          ->mutable_route()
+                                          ->mutable_weighted_clusters()
+                                          ->add_clusters();
```
Comment (adisuissa) [inline test/integration/weighted_cluster_integration_test.cc:77]: Haven't the cluster already been added above?
Why does it need to be added again?

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
Here's a first pass.

File: source/common/router/config_impl.cc [hunk 2]
```diff
@@ -999,21 +1035,31 @@ void RouteEntryImplBase::validateClusters(
...
-        throw EnvoyException(
-            fmt::format("route: unknown weighted cluster '{}'", cluster->clusterName()));
+      if (!cluster_info_maps.hasCluster(cluster->clusterName()) &&
+          cluster->clusterHeaderName().get().empty()) {
+        if (!cluster->clusterName().empty()) {
+          throw EnvoyException(
+              fmt::format("route: unknown weighted cluster '{}'", cluster->clusterName()));
+        } else {
+          throw EnvoyException("route: unknown weighted cluster with no cluster_header field");
+        }
```
Comment (snowp) [inline source/common/router/config_impl.cc:1062]: This block is really hard to follow, do you think you can simplify the checks here a bit? Maybe breaking it into several if checks or using local variables to make it easier to follow?

File: test/integration/base_integration_test.h [hunk 3]
```diff
@@ -69,7 +70,7 @@ class BaseIntegrationTest : protected Logger::Loggable<Logger::Id::testing> {
   // configuration generated in ConfigHelper::finalize.
   void skipPortUsageValidation() { config_helper_.skipPortUsageValidation(); }
   // Make test more deterministic by using a fixed RNG value.
-  void setDeterministic() { deterministic_ = true; }
+  void setDeterministicValue(uint64_t value) { deterministic_value_ = value; }
```
Comment (snowp) [inline test/integration/base_integration_test.h:73]: I'd probably have kept this named the same thing and use a default parameter, since it seems like most call sites don't care about setting a specific parameter. This should avoid some of the churn in existing tests

File: test/integration/filters/repick_cluster_filter.h [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace RepickClusterFilter {
+
+// The cluster names and cluster headers here need to be correlated correctly with
+// weighted_cluster_integration_test or any other end users, to make sure that header modifications
+// have been done on the correct target. So they are declared and defined here.
+inline constexpr absl::string_view ClusterName = "cluster_%d";
+inline constexpr absl::string_view ClusterHeaderName = "cluster_header_%d";
```
Comment (snowp) [inline test/integration/filters/repick_cluster_filter.h:12]: This feels like it might be unnecessarily complicate these tests, how bad is it to just hard code the header modifications that it does?

File: test/integration/filters/repick_cluster_filter.cc [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+#include "test/extensions/filters/http/common/empty_http_filter_config.h"
+
+#include "absl/strings/str_format.h"
+
+namespace Envoy {
+namespace RepickClusterFilter {
+
+// A test filter that modifies the request header (i.e. map the cluster header
+// to cluster name), clear the route cache, then will trigger the cluster
+// re-pick at router filter later in the filter chain.
```
Comment (snowp) [inline test/integration/filters/repick_cluster_filter.cc:20]: Seems like all this filter does is insert a single header (?) and then clears out the route cache, so I think this comment can be simplified?

File: test/integration/weighted_cluster_integration_test.cc [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+TEST_F(WeightedClusterIntegrationTest, SteerTrafficToOneClusterWithHeader) {
+  const std::vector<uint64_t>& default_weights = getDefaultWeights();
+  uint64_t deterministric_value = std::accumulate(
+      default_weights.begin(), default_weights.begin() + TotalUpstreamClusterWithNameCount, 0UL);
+  setDeterministicValue(deterministric_value);
+  initializeConfig(default_weights);
+
+  // The expected destination cluster upstream is index
+  // `TotalUpstreamClusterWithNameCount`.
```
Comment (snowp) [inline test/integration/weighted_cluster_integration_test.cc:162]: I would probably have written this test has having `WeightedClusterIndex = 0` and `WeightedClusterHeaderIndex = 1`, with one cluster designated as the one that we hit via the original cluster weighting mechanism, the other one being targeted via a cluster header. I think that would make these tests a bit easier to follow instead of having to just work with generic cluster indices.

File: test/integration/server.h [hunk 7]
```diff
@@ -430,10 +430,11 @@ class IntegrationTestServer : public Logger::Loggable<Logger::Id::testing>,
   void onWorkersStarted() override {}
 
   void start(const Network::Address::IpVersion version,
-             std::function<void()> on_server_init_function, bool deterministic,
-             bool defer_listener_finalization, ProcessObjectOptRef process_object,
-             Server::FieldValidationConfig validation_config, uint32_t concurrency,
-             std::chrono::seconds drain_time, Server::DrainStrategy drain_strategy,
+             std::function<void()> on_server_init_function,
+             absl::optional<uint64_t> deterministic_value, bool defer_listener_finalization,
```
Comment (snowp) [inline test/integration/server.h:434]: @alyssawilk Mind giving this change to propagate a deterministic value through the integration test framework a look?

Comment (alyssawilk) [reply test/integration/server.h:434]: Looks solid!  Lots of plumbing but thankfully there was that argument to replace so there wasn't a need to track APIs down from scratch :-)

I could go either way on more-accurate-naming vs churn so I'll leave that up to you.

File: test/integration/weighted_cluster_integration_test.cc [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+          }
+
+          hcm.mutable_route_config()
+              ->mutable_virtual_hosts(0)
+              ->mutable_routes(0)
+              ->mutable_route()
+              ->mutable_weighted_clusters()
+              ->mutable_total_weight()
+              ->set_value(std::accumulate(weights.begin(), weights.end(), 0UL));
+        });
```
[the remark this answers is not in the document]
Comment (tyxia) [reply test/integration/weighted_cluster_integration_test.cc:106]: I did some quick look/search, but i am not sure if fmt::format is a good choice here because:
1. It seems that templated (via fmt::format) yaml string will require more work to spell out virtual_hosts configuration (e.g. name, domains, match etc, which are default in integration test framework.)
2. I searched `addConfigModifier` usage, all the occurrence seems are using callback method.

I used yaml string in unit test (i.e. config_impl_test) where it has simple configuration. But I am not familiar with `templated (via fmt::format) yaml string` and would like to learn your thoughts. Thanks!

File: test/integration/filters/repick_cluster_filter.h (hunk 4 above)
Comment (tyxia) [reply test/integration/filters/repick_cluster_filter.h:12]: Doing this can enable generic testing approach (i.e. random combination of cluster_header + cluster_name).
I lay out some rational behind the generic approach. We can discuss further there. So I just resolve/close the conversation here.

File: test/integration/server.h (hunk 7 above)
Comment (tyxia) [reply test/integration/server.h:434]: I assume this part is LGed then :) 

Please do reopen if anything needed to be changed. Thanks!

**Review by tyxia — COMMENTED** [review]
Thanks for the thorough review!

File: test/integration/weighted_cluster_integration_test.cc (hunk 6 above)
Comment (snowp) [reply test/integration/weighted_cluster_integration_test.cc:168]: While generic code can be great, I'm hesitant to merge in generic code that is only used with one configuration. While generic code is great, it does bring with it complexity that makes this code harder to understand for everyone reading it going forward. I would suggest simplifying the test filter to only doing a single, simple thing and then have the integration test itself use well named clusters. 

It's great that this pattern let you verify things locally, but I don't think we should be checking in code that is more complex than necessary for what is actually run in CI

Comment (tyxia) [reply test/integration/weighted_cluster_integration_test.cc]: Thanks for laying out the thoughts!

I took a closer look at code.  After changing the `random number of cluster` to `fixed number (i.e. 2) cluster`, I feel most code will be same except few things: 1) we can have fixed cluster string name e.g. `cluster_header_0`  rather than `absl::StrFormat(ClusterHeaderName, i)` to construct to name 2) We have simpler pattern of clusters (just 2 clusters) and avoid some math calculation for cluster array.

I think we will still have some global variable values rather than using magic number `1` or `2` multiple places as that might not be good for readability.  

So just my 2 cents, I don't feel there are lots of code/things would be simplified by changing to fixed number cluster . Also, the flexibility and configurability outweigh the things could be simplified here. More comments may helps here

I added some comments, tweak some variable names and add  local variables in the hope of clarifying the code. Please let me know your thoughts. Thanks

Comment (tyxia) [reply test/integration/weighted_cluster_integration_test.cc]: Per offline discussion, i tweaked the variable names and code comments.

I also tweaked the cluster array config a bit: change from 2 (1 `name`+ 1 `cluster_header`) to (1 `name` + 2 `cluster_header`) .

PTAL! Thanks.

**Comment by tyxia** [comment]
/assign @yanavlasov 

Per offline conversation with @snowp , due to his upcoming vacation he would like to hand the review to @yanavlasov.

File: test/integration/weighted_cluster_integration_test.cc [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+        ConfigHelper::setHttp2(*cluster);
+      }
+
+      auto* cluster_with_name = bootstrap.mutable_static_resources()->mutable_clusters(0);
+      cluster_with_name->set_name(
+          absl::StrFormat(Envoy::RepickClusterFilter::ClusterNamePrefix, 0));
+      ConfigHelper::setHttp2(*cluster_with_name);
+
+      // Add the custom filter.
+      config_helper_.addFilter("name: repick-cluster-filter");
```
Comment (yanavlasov) [inline test/integration/weighted_cluster_integration_test.cc:65]: Please move this outside of the `addConfigModifier` lambda.

File: test/integration/weighted_cluster_integration_test.cc [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+  }
+
+  void initializeConfig(const std::vector<uint64_t>& weights) {
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      // It starts from 1 here because the first cluster is existing cluster configured by
+      // `mutable_clusters` below.
+      for (int i = 1; i < ClusterCount; ++i) {
+        auto* cluster = bootstrap.mutable_static_resources()->add_clusters();
+        cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+        cluster->mutable_load_assignment()
```
Comment (yanavlasov) [inline test/integration/weighted_cluster_integration_test.cc:48]: You do not need to set the address or protocol options. Framework will take care of this. All you need is this:
```
bootstrap.mutable_static_resources()->mutable_clusters()->Add()->MergeFrom(
    *bootstrap.mutable_static_resources()->mutable_clusters(0));
bootstrap.mutable_static_resources()->mutable_clusters(1)->set_name("cluster_1");
```

File: test/integration/weighted_cluster_integration_test.cc [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+public:
+  WeightedClusterIntegrationTest()
+      : HttpIntegrationTest(Http::CodecClient::Type::HTTP2, Network::Address::IpVersion::v6) {
+    default_weights_.reserve(ClusterCount);
+    std::fill_n(std::back_inserter(default_weights_), ClusterNameClusterCount, 20);
+    std::fill_n(std::back_inserter(default_weights_),
+                Envoy::RepickClusterFilter::ClusterHeaderClusterCount, 30);
+  }
+
+  void createUpstreams() override {
```
Comment (yanavlasov) [inline test/integration/weighted_cluster_integration_test.cc:33]: Looking at it more, you cam remove the `createUpstreams()` and just call `setUpstreamCount` in the constructor.

File: test/integration/weighted_cluster_integration_test.cc [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+        [&weights](
+            envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+                hcm) {
+          // Add the clusters with `name` specified.
+          for (int i = 0; i < ClusterNameClusterCount; ++i) {
+            auto* cluster_with_name = hcm.mutable_route_config()
+                                          ->mutable_virtual_hosts(0)
+                                          ->mutable_routes(0)
+                                          ->mutable_route()
+                                          ->mutable_weighted_clusters()
```
Comment (yanavlasov) [inline test/integration/weighted_cluster_integration_test.cc:79]: You can first store the pointer from the `mutable_weighted_clusters()` and then call `add_cluster()` on it. It will remove duplication of this long chain of calls.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

**Review by tyxia — COMMENTED** [review]
Thanks for the review!

File: test/integration/weighted_cluster_integration_test.cc (hunk 11 above)
Comment (yanavlasov) [reply test/integration/weighted_cluster_integration_test.cc:22]: The framework will create upstreams for you. This function just duplicates the code in the base class.

**Review by yanavlasov — APPROVED** [review]
You can reduce it a bit more by removing the createUpstreams method. Other than that looks good.

File: test/integration/weighted_cluster_integration_test.cc (hunk 11 above)
Comment (tyxia) [reply test/integration/weighted_cluster_integration_test.cc:22]: I took a closer look at this and tried it in several places. I don't think just `setUpstreamCount` will work: 
`createUpstreams()` here will invoke `addFakeUpstream` which will call this version of `FakeUpstream` constructor `FakeUpstream(Network::Test::createRawBufferSocketFactory(), makeListenSocket(config, makeAddress(port, version)), config)` via  `std::make_unique<FakeUpstream>(0, version_, config)`

However, the  `createUpstreams()` in the framework which is `BaseIntegrationTest::createUpstreams()` will invoke different constructor of `FakeUpstream(std::move(transport_socket_factory), makeListenSocket(config, address), config)` via `new FakeUpstream(std::move(factory), endpoint, upstreamConfig()`. 

I didn't investigate further because `createUpstreams()` has been used in many integration tests (i.e. I feel such change doesn't introduce extra complexity) but i feel the way of setting of the ListenSocket address (mentioned above) is the key here.  Also, I do feel the framework might need some clean-up/refactor to avoid confusion like this(which can be a TODO).

**Comment by tyxia** [comment]
> You can reduce it a bit more by removing the createUpstreams method. Other than that looks good.

I took a closer look at this and tried it in several places. I don't think just `setUpstreamCount` will work:
createUpstreams() here will invoke `addFakeUpstream` which will call this version of FakeUpstream constructor `FakeUpstream(Network::Test::createRawBufferSocketFactory(), makeListenSocket(config, makeAddress(port, version)), config)` via `std::make_unique<FakeUpstream>(0, version_, config)`

However, the `createUpstreams()` in the framework which is `BaseIntegrationTest::createUpstreams()` will invoke different constructor of `FakeUpstream(std::move(transport_socket_factory), makeListenSocket(config, address), config)` via `new FakeUpstream(std::move(factory), endpoint, upstreamConfig()`.

I didn't investigate further because createUpstreams() has been used in many integration tests (i.e. I feel such change doesn't introduce extra complexity) but i feel the way of setting of the ListenSocket address (mentioned above) is the key here. Also, I do feel the framework might need some clean-up/refactor to avoid confusion like this(which can be a TODO).
